_posts/2020-08-24-native-loader-net5.md

```diff
@@ -0,0 +1,38 @@
+
+The managed loading changes are fairly clear and well documented, but unmanaged library loading has changed in numerous ways, some of them far more subtle.
+
+First off, Mono now follows [this algorithm](https://docs.microsoft.com/en-us/dotnet/core/dependency-loading/loading-unmanaged) for P/Invoke resolution. Notably absent here is DllMap's XML config files, which have been disabled entirely in .NET 5. The DllMap embedding APIs are also disabled by default on desktop platforms, but can be enabled by building with `--enable-minimal=dllmap`.
+
+In place of DllMap, users are encouraged to utilize the [NativeLibrary](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.nativelibrary?view=netcore-3.1) resolution APIs, which are set in managed code, and the [runtime hosting properties](https://docs.microsoft.com/en-us/dotnet/core/tutorials/netcore-hosting#step-3---prepare-runtime-properties), which are set by embedders with the `monovm_initialize` function. If these are insufficient for your use case, please [tell us about it](https://github.com/dotnet/runtime/issues/37213)! We're always looking to improve our interop functionality, and in particular with .NET 6 will be evaluating `NativeLibrary`, so please speak up.
+
+A more subtle, yet no less impactful change is that native library loading now defaults to `RTLD_LOCAL` to be consistent with CoreCLR and Windows, as opposed to our historical behavior of `RTLD_GLOBAL`. This change prevents symbol collision, and will both break and enable various scenarios/libraries. What this means in practice is that on Unix-like platforms, libraries are longer loaded into a single global namespace and when looking up symbols, the specified library must be correctly specified.
+
+Historically, in Mono on Linux it was possible to load library `foo` containing symbol `bar`, and then invoke `bar` with a P/Invoke like so: 
+```
+[DllImport("asdf")]
+public static extern int bar();
+```
+
+This will no longer work. For that P/Invoke to function correctly, the attribute would need to use the correct library name: `[DllImport("foo")]`. A lot of code in the wild that was using incorrect library names will need to be updated. However, this means that when loading two libraries containing the same symbol name, there is no longer a conflict.
+
+There have been some embedding API changes as part of this. `MONO_DL_MASK` is no longer a full mask, as `MONO_DL_GLOBAL` has been introduced to specify `RTLD_GLOBAL`. If both `MONO_DL_LOCAL` and `MONO_DL_GLOBAL`, are set, Mono will use local. See mono/utils/mono-dl-fallback.h for more info.
```

Comment:
Interesting choice here. Why not imply the absence of GLOBAL to be local? That way there would be no way to specify both.

Reply:
Do you mean the absence of `LOCAL`? The embedding API historically has defaulted to `GLOBAL`, but that's what I've changed here; it now defaults to `LOCAL`. If I don't extend it, there will be no way to override that.

Reply:
For what it's worth, I considered leaving the embedding API to default to `GLOBAL` and just changing the `PInvoke` behavior, but ultimately decided this would be oddly confusing/inconsistent and not push people towards just using `LOCAL` by default.